ru/lessons/advanced/metaprogramming.md

```diff
@@ -0,0 +1,238 @@
+
+```elixir
+defmodule OurMacro do
+  defmacro unless(expr, do: block) do
+    quote do
+      if !unquote(expr), do: unquote(block)
+    end
+  end
+end
+```
+
+Давайте подключим этот модуль и попробуем использовать этот макрос:
+
+```elixir
+iex> require OurMacro
+nil
+iex> OurMacro.unless true, do: "Hi"
+nil
+iex> OurMacro.unless false, do: "Hi"
+"Hi"
+```
+
+Так как макрос заменяет код в самом приложении, мы можем контролировать что и когда компилируется. Отличным примером является модуль `Logger`. Когда логирование отключено, код не подключается и результирующее приложение не содержит никаких отсылок к коду логирования. Этим Elixir отличается от других языков, в которых будут накладные расходы на вызовы пустых функций.
+
+Для демонстрации этого давайте сделаем простой логгер, который может быть включен или выключен.
+
+```elixir
+defmodule Logger do
+  defmacro log(msg) do
+    if Application.get_env(:logger, :enabled) do
+      quote do
+        IO.puts("Логируемое сообщение: #{unquote(msg)}")
+      end
+    end
+  end
+end
+
+defmodule Example do
+  require Logger
+
+  def test do
+    Logger.log("Это запись в журнале")
+  end
+end
+```
+
+Когда логирование включено, наша функция `test` сгенерирует приблизительно такой код:
+
+```elixir
+def test do
+  IO.puts("Logged message: #{"Это запись в журнале"}")
+end
+```
+
+А когда логирование будет отключено - вот такой:
+
+```elixir
+def test do
+end
+```
+
+### Закрытые макросы
+
+Хоть это и не частый случай, Elixir поддерживает закрытые макросы. Закрытый макрос определяется вызовом `defmacrop` и может быть вызван только из того модуля, в котором он был определен. Закрытые макросы должны быть определены раньше чем будут вызваны кодом.
+
+### Гигиена при использовании макросов
+
+Как макросы взаимодействуют с контекстом вызова называется гигиеной макросов. По умолчанию, макрос в Elixir гигиеничен и не будет конфликтовать с контекстом:
+
+```elixir
+defmodule Example do
+  defmacro hygienic do
+    quote do: val = -1
+  end
+end
+
+iex> require Example
+nil
+iex> val = 42
+42
+iex> Example.hygienic
+-1
+iex> val
+42
+```
+
+Но что если мы хотим изменять значение `val`? Для того чтобы обозначить переменную негигиеничной мы можем использовать `var!/2`. Давайте изменим пример с использованием этого трюка:
+
+```elixir
+defmodule Example do
+  defmacro hygienic do
+    quote do: val = -1
+  end
+
+  defmacro unhygienic do
+    quote do: var!(val) = -1
+  end
+end
+```
+
+И посмотрим как он теперь взаимодействует с контекстом:
+
+```elixir
+iex> require Example
+nil
+iex> val = 42
+42
+iex> Example.hygienic
+-1
+iex> val
+42
+iex> Example.unhygienic
+-1
+iex> val
+-1
+```
+
+Включая `var!/2` в макрос мы изменили значение `val` не передавая его внутрь макроса. Использование таких макросов должно быть минимальным, ведь таким образом мы увеличиваем вероятность конфликта имен.
```

Comment:
Several commas are missing here. The first sentence should look like the following: 
>Включая `var!/2` в макрос, мы изменили значение `val`, не передавая его внутрь макроса.